File: C#/第一部分 C#语言/第八章 引用与内存模型/第三节 链表结构/linked_list.cs
// 由bhw设计
/*
第8章 第3节 链表结构

一、原理
1) 链表由节点构成，每个节点保存“值 + 下一个节点引用”。
2) 相比数组，链表插入删除更灵活，但随机访问慢。
3) 信息学里常用于理解指针思想和动态结构。

二、习题
1) 在头部插入一个新节点。
2) 删除值为 x 的第一个节点。

三、答案与讲解
1) newNode.Next = head; head = newNode;
2) 需要处理删除头节点和中间节点两种情况。
*/
using System; // 输出库

class Node // 定义链表节点类
{
    public int Value; // 节点值
    public Node? Next; // 指向下一个节点
}

class Program
{
    static void Main()
    {
        Node a = new Node { Value = 1 }; // 创建节点 a
        Node b = new Node { Value = 2 }; // 创建节点 b
        Node c = new Node { Value = 3 }; // 创建节点 c
        a.Next = b; // a 指向 b
        b.Next = c; // b 指向 c

        Node? p = a; // 定义遍历指针
        while (p != null) // 遍历到末尾为止
        {
            Console.Write(p.Value + " "); // 输出当前节点值
            p = p.Next; // 移动到下一个节点
        }
        Console.WriteLine(); // 换行
    }
}
// 由bhw设计

File: C/第一部分 C语言/第八章 指针及其应用/第六节 链表结构/linked_list.c
// 由bhw设计
/*
第8章 第6节 链表结构
核心点:
1) 链表节点包含数据域和指针域。
2) 通过next把节点串起来。
3) 插入删除无需整体搬移元素。

习题:
1) 实现头插法创建链表。
2) 删除值为x的第一个节点。

答案与讲解:
1) 新节点next指向原头, 再更新头指针。
2) 遍历时记录前驱节点再断链。
*/
#include <stdio.h> // 输出函数

struct Node { // 定义链表节点结构体
    int val; // 节点值
    struct Node *next; // 指向下一个节点
}; // 结构体定义结束

int main(void) { // 主函数开始
    struct Node a = {1, NULL}; // 定义节点a
    struct Node b = {2, NULL}; // 定义节点b
    struct Node c = {3, NULL}; // 定义节点c
    a.next = &b; // 连接a到b
    b.next = &c; // 连接b到c
    for (struct Node *p = &a; p != NULL; p = p->next) { // 遍历链表
        printf("%d ", p->val); // 输出当前节点值
    } // 遍历结束
    printf("\n"); // 输出换行
    return 0; // 正常结束
} // main结束
// 由bhw设计
